pkg/kubelet/kubelet_pods.go

```diff
@@ -977,6 +977,11 @@ func (kl *Kubelet) PodResourcesAreReclaimed(pod *v1.Pod, status v1.PodStatus) bo
 	// physically deleted.
 
 	klog.V(3).InfoS("Pod is terminated and all resources are reclaimed", "pod", klog.KObj(pod))
+
+	if !kl.podWorkers.IsPodTerminationFinished(pod.UID) {
```

Comment:
This is a pretty dramatic change to the logic of deletion.  Now we can't delete the pod object (meaning stateful sets block longer) until everything in SyncTerminatedPod has completed, whereas before we could.

In *practice* everything in SyncTerminatedPod is already covered here if we keep this check.  This new check you add obviates CouldHaveRunningContainers and countRunningContainerStatus.  podVolumesExist would have to be moved to SyncTerminatedPod.  Cgroup teardown already blocks.

The real slow part is volume teardown - that is *not* done in SyncTerminatedPod today (just detach), but it's not clear that we shouldn't be tearing down the volume anyway before we indicate terminal phase.  I'll need to get someone from sig storage to weigh in on that, but it might be ok.

So to summarize - if we do this change, then this whole method goes away and instead status manager just checks podWorkers.IsPodTerminationFinished(), and we add the podVolumesExist check in SyncTerminatedPods() right below unmount.  Hold briefly for that.

Reply:
cc @gnufied to double check my conclusions

Reply:
Would you expect that disk directory cleanup to be fast?

Reply:
Yes, after the unmount operation success, it's just a matter of rmdir (assuming there aren't some other weird bugs).

Reply:
It seems like something that might require further exploring and doesn't seem like a regression. Do you think it could be decoupled and taken as a follow up by folks from sig-node / sig-storage?

Reply:
Ok, then I'm going to guide us to reduce overall complexity.

@mimowo:

1. PodResourcesAreReclaimed() and podResourcesAreReclaimed() should be removed
2. Your new pod worker method I would recommend calling `ShouldPodBeFinished(uid) bool`
3. That method should `return p.podsSynced` at the end
4. I would recommend changing the eviction manager podCleanupFunc signature to match `ShouldPodBeFinished`
5. Status manager should only consult `podIsTerminal` (from the statusmanager result) instead of calling this (and you should remove the old method we don't use anymore from what status manager needs)
6. We should update the SyncPod interface godoc for SyncTerminatedPod to describe that until the method exits the pod cannot be deleted.

That should dramatically simplify code.

As part of reviewing this, I realized that status manager holding pod status is wrong, that's part of "problem 9" in the kubelet state deck https://docs.google.com/presentation/d/1wbPFANXVbk-e-5BvVHaZScqoqJsfcU_FQVmG-N7pv3c/edit#slide=id.g1de8a1ca1a4_0_1865 but I'll have to review more.

Reply:
Thanks I will follow up on the suggestions, I think it makes sense. However, I think `PodCouldHaveRunningContainers` still might be needed, because the status update for SyncTerminatedPod (with podIsTerminal) might get lost if the `podStatusChannel` channel is full, see: https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/status/status_manager.go#L663-L668. In that case the correct status is set by `SyncTerminatedPod`, but the actual deletion is progressed by syncPod called from syncBatch. In that case `podIsTerminal=false`, but the fallback to `PodCouldHaveRunningContainers` guarantees deletion.

Reply:
> because the status update for SyncTerminatedPod (with podIsTerminal) might get lost if the podStatusChannel channel is full, see:

Good point, that channel will be removed in the future (that's what problem 9 describes), when that gets removed we can fix this then.

Reply:
Hrm... this is ugly - if we can fail to put podsIsTerminal in the status manager, then we can't rely on it being there to finish deletion.

Let me look and see how difficult is is to remove just the channel.

Reply:
> Hrm... this is ugly - if we can fail to put podsIsTerminal in the status manager, then we can't rely on it being there to finish deletion.

I've tested that by inducing random drop of the items in the channel. It works fine, because `SyncTerminatedPod` leaves the correct status in the cache. Then the status is picked up by syncPod from syncBatch and delivered, only then the status is deleted, but indeed it is complex.

Reply:
Couple of remaining questions:
1 and 5. Ok, so if we don't drop `PodResourcesAreReclaimed` - due to the `podStatusChannel`. I guess relatively easily we could rename it to `PodIsTerminal`?
4. Could you point more specifically. Could this be deferred for later PRs, preferrably owned by sig-node?
7. Do we have some ready-to-use function for that? Is this something that could be deferred for later PRs, preferrably owner by sig-node?

Reply:
Ok, I fixed the status manager in https://github.com/kubernetes/kubernetes/pull/116615 so that there is no longer a queue (as a side effect status updates propagate faster and we write fewer overall).  You can pick that into your PR once that goes green.

Reply:
What if the status update for `syncTerminatedPods` fails on the PATCH request, will `syncTerminatedPod` be repeated, or we still need to rely on syncBatch to re-attempt applying the patch and deleting the pod?

Reply:
`needsUpdate` should be true (canBeDeleted() should be true), so the next syncBatch will retry the delete and succeed.  The status cache should have `podIsTerminal=true` but that would be a test we would add.  Let me see if we already have a status manager test for when the last patch fails.

Reply:
For 4 this should be pretty straightforward, I can help tomorrow.  7 we should get the right code tomorrow from sig-node.

Reply:
Thanks, update from me: 
1. I have cherry-picked your PR and it appears to work in combination with mine (I'm still to carry more complex testing).
2. I've dropped the `PodResourcesAreReclaimed` and replaced its use with `PodTerminationIsFinished` which calls podWorkers. This is still needed to ensure `canBeDeleted` returns true in case the status update for `SyncTerminatedPod` failed and `syncBatch` needs to continue
3. I renamed `podIsTerminal` to `podTerminationIsFinished` on the path for `SyncTerminatedPod`. The reason is that there is a slight difference in semantics actually. `podIsTerminal=true` is used in `generateAPIPodStatus` as soon as the containers are killed in `SyncTerminatingPod` (and at the beginning of `SyncTerminatedPod`), but not necessarily when the cleanup is done, so we still want to hold the deletion. I think we need the distintion, otherwise we could delete a pod too early. On the other hand, we need to generate the status as soon as the containers are killed, but before they are removed (thus at the end `SyncTerminatingPod`). 

I consider that alternatively we could keep the `podTerminationFinished` is kept inside `versionsedPodStatus` and once true, it never flips back to `false` when generating a new version of the version status. This would eliminate the need for the `PodTerminationIsFinished` function entirely. The downside is that we need to pass the new boolean from every `SetPodStatus` and there are a few places it is called from. WDYT? 

Let me know, I will try to make it work with extending `podVersionedStatus` in the meanwhile and see if this works.

Reply:
I've pushed the commit with using `versionedPodStatus` to embed `podTerminationFinished`. This allows to completely drop `PodResourcesAreReclaimed` and  `PodTerminationIsFinished`. PTAL.

Reply:
Ok, in my latest commit I've gone possible too far - I introduce podIsTerminal along with podIsFinished. Then, I drop entirely the podDeletionSafetyProvider. The `podIsTerminal` is used instead of CouldHaveRunningContainers. The reason is that there is an issue with CouldHaveRunningContainers - it is checked async of computing the status. Thus, it was possible that the status was computed while containers where running, but at the moment of mergePodStatus the containers aren't running. Thus, we could get Failed phase (for example in case of eviction), but all containers would still be running in the status. This isn't critical, as I mostly care about the end phase, but I thought might be nice to solve while on that. If you think this is a good idea I will clean it up entirely, otherwise revert.

Reply:
tested new tests with simulating errors on applying the patch (to test the syncBatch) and they pass:
```golang
	mergedStatus := mergePodStatus(pod.Status, status.status, !status.podIsTerminal)
	if rand.Intn(3) > 0 {
		klog.InfoS("Failed to apply the Patch Status update on random!!!")
		return
	}
```
Going to perform more stability tests in a loop soon.

Reply:
I think I understand the complications, but don't see an easy way out, reverting to the point when it worked (with `PodDeletionSafetyProvider`. I guess the rare issue with not accurate pod containers might be ticketed for later.

Reply:
Resolving as `CouldHaveRunningContainers` is gone, and the wait for volume path cleanup is now done in SyncTerminatedPod.